core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java

```diff
@@ -117,28 +80,68 @@ public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput sliceIn
         return new VariableWidthBlock(0, positionCount, slice, offsets, valueIsNull);
     }
 
-    private static void computeOffsetsFromLengths(int[] offsets, boolean[] valueIsNull, int lengthIndex)
+    private static void writeOffsetsWithNullsCompacted(SliceOutput sliceOutput, int[] rawOffsets, @Nullable boolean[] valueIsNull, int baseOffset, int positionCount)
+    {
+        checkFromIndexSize(baseOffset, positionCount + 1, rawOffsets.length);
+
+        int startingOffset = rawOffsets[baseOffset];
+        if (valueIsNull == null && startingOffset == 0) {
+            // No translation of offsets required, write the range of raw offsets directly to the output
+            sliceOutput
+                    .appendInt(positionCount)
+                    .writeInts(rawOffsets, baseOffset + 1, positionCount);
+        }
+        else if (valueIsNull == null) {
+            // Subtract starting offset from each offset value to translate them to start from zero, no null suppression required
+            int[] nonNullEndingOffsets = new int[positionCount];
+            for (int i = 0; i < nonNullEndingOffsets.length; i++) {
+                nonNullEndingOffsets[i] = rawOffsets[i + baseOffset + 1] - startingOffset;
+            }
+            sliceOutput
+                    .appendInt(nonNullEndingOffsets.length)
+                    .writeInts(nonNullEndingOffsets, 0, nonNullEndingOffsets.length);
+        }
+        else {
+            // Translate offsets and suppress null values from the output
+            int[] nonNullEndingOffsets = new int[positionCount];
+            int nonNullEndingOffsetCount = 0;
+            for (int i = 0; i < positionCount; i++) {
+                nonNullEndingOffsets[nonNullEndingOffsetCount] = rawOffsets[i + baseOffset + 1] - startingOffset;
+                nonNullEndingOffsetCount += valueIsNull[i + baseOffset] ? 0 : 1;
+            }
+            sliceOutput
+                    .appendInt(nonNullEndingOffsetCount)
+                    .writeInts(nonNullEndingOffsets, 0, nonNullEndingOffsetCount);
+        }
+    }
+
+    private static int[] readOffsetsWithNullsCompacted(SliceInput sliceInput, @Nullable boolean[] valueIsNull, int positionCount)
     {
-        if (lengthIndex < 0 || lengthIndex > offsets.length) {
-            throw new IllegalArgumentException(format("Invalid lengthIndex %s for offsets %s", lengthIndex, offsets.length));
+        if (valueIsNull != null && valueIsNull.length != positionCount) {
+            throw new IllegalArgumentException(format("valueIsNull length must match positionCount, found %s <> %s", valueIsNull.length, positionCount));
+        }
+        int nonNullEndingOffsets = sliceInput.readInt();
```

Comment:
The `Ending` part in `nonNullEndingOffsets` seems unnecessary. You could mention it once in a comment somewhere, but I think it is understood that VariableWdithBlock and related things work only on "end" offsets.